### PR TITLE
#1208 Support relative context URI starting with "$metadata"

### LINKED ISSUE #1208: ODataLib Client fails to deserialize payload with @odata.context relative URL.
*When JSON Lite payload contains @odata.context annotation with relative URL instead of absolute URL OData .NET client fails to deserialize response.*

### Assemblies affected

*OData .Net lib 7.5.0.*

### Reproduce steps

*Read response with relative URL in @odata.context annotation. (Use Fiddler Autoresponder?).*

### Expected result

*Client should deserialize response as usual.*

### Actual result

*Client fails with following error:
Microsoft.OData.ODataException: 'The top level context URL '$metadata#FooBar' should be an absolute Uri.'.*

### Additional detail

*According to [OData V4 spec](http://docs.oasis-open.org/odata/odata-json-format/v4.01/cs01/odata-json-format-v4.01-cs01.html#_Toc499720585):
The context annotation returns the context URL (see [OData-Protocol]) for the payload. This URL can be absolute or **relative**.
This output comes from Java Olingo implementation and we are blocked by this issue right now.*

### PR DESCRIPTION
**Pull request by nobyk**
### Issues

*This pull request fixes #1208.*

### Description

*Relative context URI could start with "$metadata". This PR covers the case in addition to the fix of [#2132: Relative URIs for @oData.context are not supported.](https://github.com/OData/odata.net/pull/2216)*

### Checklist (Uncheck if it is not completed)

- [X] *Test cases added*
- [X] *Build and test with one-click build and test script passed*

### CODE REVIEW

**Comment by nobyk** [comment]
This PR won't cover related entity addressing, when OData service returns `"@odata.context": "../$metadata#Supplier"` for a request `https://host/service/Products(1)/Supplier`.

**Comment by nobyk** [comment]
If PR https://github.com/OData/odata.net/pull/1211 will be merged soon, it's OK to close this PR without merge. But the PR is open for a while and PR https://github.com/OData/odata.net/pull/2216 was merged in 7.10.0, so I would like to merge this PR as small and fast resolution.

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightDeserializer.cs [hunk 1]
```diff
@@ -249,10 +249,12 @@ internal static bool TryParsePropertyAnnotation(string propertyAnnotationName, o
             {
                 // If the Base uri string is http://odata.org/test
                 // The MetadataDocumentUri will be http://odata.org/test/$metadata
-                // If the contextUriAnnotation value is Customers(1)/Name
+                // If the contextUriAnnotation value is Customers(1)/Name or $metadata#Customers(1)/Name
                 // The generated context uri will be http://odata.org/test/$metadata#Customers(1)/Name
                 ODataUri oDataUri = new ODataUri() { ServiceRoot = this.BaseUri };
-                contextUriAnnotationValue = oDataUri.MetadataDocumentUri.ToString() + ODataConstants.ContextUriFragmentIndicator + contextUriAnnotationValue;
+                contextUriAnnotationValue = contextUriAnnotationValue.StartsWith("$metadata", StringComparison.OrdinalIgnoreCase)
```
Comment (KenitoInc) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightDeserializer.cs:255]:
```suggestion
                contextUriAnnotationValue = contextUriAnnotationValue.StartsWith("$metadata#", StringComparison.OrdinalIgnoreCase)
```

**Review by KenitoInc — APPROVED** [review]

**Comment by KenitoInc** [comment]
@nobyk 
Thanks for the PR. I have requested one minor change

**Comment by nobyk** [comment]
@KenitoInc, Thank you for your suggestion. I've applied it.

**Review by gathogojr — APPROVED** [review]
:shipit:

**Review by habbes — APPROVED** [review]

**Comment by KenitoInc** [comment]
@nobyk Thanks for this PR

**Pull request merged**